katdal/applycal.py

```diff
@@ -394,10 +451,15 @@ def calc_correction(chunks, cache, corrprods, cal_products):
                         data[j] = v
             else:
                 data = sensor
-            products[product].append(data)
-    params = CorrectionParams(inputs, input1_index, input2_index, products)
-    name = 'corrections[{}]'.format(','.join(cal_products))
-    return da.map_blocks(_correction_block, chunks=chunks, dtype=np.complex64, params=params)
+            corrections_per_product.append(data)
+        else:
+            corrections[cal_product] = corrections_per_product
+    if not corrections:
+        return None
+    params = CorrectionParams(inputs, input1_index, input2_index, corrections)
+    name = 'corrections[{}]'.format(','.join(corrections.keys()))
```

Comment:
Maybe worth sorting keys to give reproducible ordering (particularly since 2.7, 3.5 and 3.6+ will all do different things)?

Reply:
The name is only for display / debugging purposes, but this sounds reasonable.

Reply:
It's not just for debugging purposes. It forms part of the dask key, which dask uses to decide when two things are the same and don't need to be recomputed. I don't know enough about dask.distributed to know if this is actually a concern in practice, but I think it's at least possible for two processes in a dask cluster to get different orders and hence do unnecessary recomputation.

Reply:
Good point, forgot about that!